src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,986 @@
+                foreach (var identifierName in queryExpressionProcessingInfo.IdentifierNames)
+                {
+                    // Identifier from the foreach can already be in scope of the foreach statement.
+                    if (forEachStatement.Identifier.ValueText != identifierName)
+                    {
+                        if (_semanticFacts.GenerateUniqueName(
+                                _semanticModel,
+                                location: forEachStatement.Statement,
+                                containerOpt: forEachStatement.Statement,
+                                baseName: identifierName,
+                                usedNames: Enumerable.Empty<string>(),
+                                _cancellationToken).ValueText != identifierName)
+                        {
+                            documentUpdateInfo = default;
+                            return false;
+                        }
+                    }
+                }
+
+                // If query does not contains identifier with the same name as declared in the foreach,
+                // declare this identifier in the body.
+                if (!queryExpressionProcessingInfo.ContainsIdentifier(forEachStatement.Identifier))
+                {
+                    documentUpdateInfo = ConvertIfInToForeachWithExtraVariableDeclaration(forEachStatement, queryExpressionProcessingInfo);
+                    return true;
+                }
+                else
+                {
+                    // The last select expression in the query returns this identifier:
+                    // foreach(var thisIdentifier in from ....... select thisIdentifier)
+                    // if thisIdentifier in foreach is var, it is OK
+                    // if a type is specified for thisIdentifier in forEach, check that the type is the same as in the select expression
+                    // foreach(MyType thisIdentifier in from ....... from MyType thisIdentifier ... select thisIdentifier)
+
+                    // The last clause in query stack must be SelectClauseSyntax.
+                    var lastSelectExpression = ((SelectClauseSyntax)queryExpressionProcessingInfo.Stack.Peek()).Expression;
+                    if (lastSelectExpression is IdentifierNameSyntax identifierName &&
+                        forEachStatement.Identifier.ValueText == identifierName.Identifier.ValueText &&
+                        _semanticModel.GetSymbolInfo(identifierName, _cancellationToken).Symbol.Kind == SymbolKind.RangeVariable)
+                    {
+                        var forEachStatementTypeSymbolType = _semanticModel.GetTypeInfo(forEachStatement.Type, _cancellationToken).Type;
+                        var lastSelectExpressionTypeInfo = _semanticModel.GetTypeInfo(lastSelectExpression, _cancellationToken);
+                        if (lastSelectExpressionTypeInfo.ConvertedType == lastSelectExpressionTypeInfo.Type &&
```

Comment:
>lastSelectExpressionTypeInfo.ConvertedType == lastSelectExpressionTypeInfo.Type  [](start = 28, length = 80)

I am not sure what motivated this change. I thing this now does a reference comparison because it compares interfaces. In general, we should not assume that equal types are reference equal, this is often false for generic types, arrays, etc

Reply:
Please see QueryInForEachWithConvertedType.
SymbolEquivalenceComparer returns true for C and for int. 

---
In reply to: [178423789](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423789) [](ancestors = 178423789)

Reply:
I thought we expect to differentiate C and int in this example. Correct?

---
In reply to: [178423950](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423950) [](ancestors = 178423950,178423789)

Reply:
> I thought we expect to differentiate C and int in this example. Correct?

Correct.
The comment was for the change that replaced ```SymbolEquivalenceComparer.Instance.Equals``` with ```==```.
But SymbolEquivalenceComparer.Instance.Equals would do what we want. I think ```==``` will fail for cases when we want the check to pass.

---
In reply to: [178423957](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423957) [](ancestors = 178423957,178423950,178423789)

Reply:
> SymbolEquivalenceComparer returns true for C and for int. 

Now I am really confused. What does ```SymbolEquivalenceComparer.Instance.Equals``` do and why we are using it for the comparison below?

---
In reply to: [178424410](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424410) [](ancestors = 178424410,178423957,178423950,178423789)

Reply:
Should we use Object.Equals to compare types instead? It looks like we are using ```SymbolEquivalenceComparer.Instance.Equals``` in several places in this file to compare types. The main point is - reference comparison is the wrong way. Object.Equals is better

---
In reply to: [178424954](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424954) [](ancestors = 178424954,178424410,178423957,178423950,178423789)